### ISSUE TITLE
Current Container Listing mechanism does not support hidden resource use case

### ISSUE THREAD

**Issue by dmitrizagidulin** [solid/solid#29]
The current mechanism to list resources in a Solid container does not support an important social media use case.

Say I'm implementing a social blog service (like LiveJournal or Facebook) that allows me to specify access control on each post: public, private (only I can see and read those posts), and friends-only.
Using the current Solid spec and implementation, I cannot simply put my Posts as resources in a `/blog/` Container, and set their corresponding ACL files. This is because when an un-authenticated user (or an authenticated user who is not on my Friends list) does a List Resources request on the `/blog/` container, they will see *all* of the posts listed on my blog, even private or friends-only ones.

In other words, as a non-friend, I do `GET dmitri.databox.me/blog/`, I will see that it contains:
* `public post 1` (which means I have read access to it, no problem)
* `private post 2` (I can see it on the list, but get a 401 / access denied when I try to request it)
* `friends-only post 3` (again, I can see it on the list but get access denied)

This is completely un-usable, in terms of being able to implement private/hidden resources in the context of a social media app.

After a discussion with @deiu, @sandhawke and @nicola, the proposal is to:

- Change the Solid spec dealing with listing of containers, do not show resources a user has no access to.
- Update the relevant implementations (ldnode, etc) to match it.

**Comment by bblfish** [comment]
Only issue I have with this is: how can a client know that if it **were** to authenticate as X then it **would** be able to see more? This problem is similar to the one raised in [issue 45: No way to find out if user is authorized to write to a container](https://github.com/solid/solid/issues/45)

**Comment by dmitrizagidulin** [comment]
I think you may be missing the point of hidden resources. The client is *not* supposed to know that if it were to authenticate it would be able to see it more. And shouldn't. As in, that's the main value of the feature.

**Comment by bblfish** [comment]
As pointed out I have a use case for hidden resource in [Storing an Apps Public Key](https://github.com/solid/solid/wiki/Storing-an-Apps-Public-Key#questions-1) and I can see the use of it in other use cases. 

But this does not mean that there are use cases where it is useful to see resources that are protected. For example 
* if I want to say there is a resource available to members of my club, but I want people to know that, so that they can join the club.
* if I have a resource that is only visible to people over a certain age, but I actually want people of that age to see it, if they can prove it.

What is actually needed is a way to state that links to a certain hidden resources should be hidden when one wishes that to be the case, and to be able to state that links to hidden resources are visible when one wants that to be the case.

So I think actually we need a vocabulary for describing views on resources, which could be part of acl.

**Comment by dmitrizagidulin** [comment]
I agree, there's definitely a need to explicitly state the two different intentions - 1) this resource is *hidden* (and so doesn't show up on lists, and accessing it yields a 404), and 2) this resource is not hidden but locked (and so yields a 401).

**Comment by melvincarvalho** [comment]
@bblfish +1 this is a new feature, and not a bug.

I would prefer this issue to be phrased in amore neutral way.

The opening premise:

> The current mechanism to list resources in a Solid container does not support an important social media use case

That's not the case is it.  The argument is that the *names* of files may be leaked, if they are put in a public container.

Of course there is another solution:  dont put the names of those in a public container.

> This is completely un-usable

I would also like to see this rephrased in more neutral language, I think it's too strong.

> Update the relevant implementations (ldnode, etc) to match it.

Also what are all the relevant implementations?  Does this not apply to all LDP next and upcoming Solid implementations?  Could we try to investigate all the systems that might need to make changes?

> After a discussion with @deiu, @sandhawke and @nicola, the proposal is to:

Do you have a pointer to this?  From what @sandhawke said in the meeting I remember, containers should either list all resources in them or none.   I may have misheard there.  It seems like this proposal actually is an architectural change.  Perhaps there's implications as to the RESTfulness etc. that may need a lot of consideration.

**Comment by ebporter** [comment]
@dmitrizagidulin Is it just that the unauthenticated user or not-connected user can see a _list_ of all the resources in the container, or that they can see the listed items themselves? Either way, it's not acceptable, I'm just trying to understand. It seems very important to support the ability to set a general permission on a container (public, for example), but allow an arbitrary number of overrides on the items in the container, including other containers, etc. For example, let's say you have a blog post in an unfinished state and you just want to save it and come back to it tomorrow. So, you want it to be private for the time being until you're ready to publish it. Anyway, there are dozens of use cases...
